crates/pico_macros/src/memo.rs

```diff
@@ -0,0 +1,134 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, FnArg, ItemFn, PatType, ReturnType};
+
+pub(crate) fn memo(_args: TokenStream, item: TokenStream) -> TokenStream {
+    let ItemFn {
+        sig,
+        vis,
+        block,
+        attrs,
+    } = parse_macro_input!(item as ItemFn);
+
+    let fn_name = sig.ident.clone().to_string();
+
+    let inputs = sig.inputs.iter().collect::<Vec<_>>();
+
+    if inputs.is_empty() {
+        return syn::Error::new_spanned(&sig, "Function must have at least one argument")
```

Comment:
nit: maybe we can specify that the first arg must be the db (though I don't know whether we can enforce that?)

Reply:
yeah, I think we can check that type is `&mut Database`

Reply:
Would that mean this would fail if you imported Database as `PicoDb` or somethig?

Reply:
(And if we *can* enforce that, hot damn, that opens up a bunch of possibilities)